### PR TITLE
Add support for MongoDB Atlas $vectorSearch vector search

### PR DESCRIPTION
**Pull request by NoahStapp**
Adds support for the `$vectorSearch` operator for MongoDBAtlasVectorSearch, which was announced at .Local London (September 26th, 2023). This change maintains breaks compatibility support for the existing `$search` operator used by the original integration (https://github.com/langchain-ai/langchain/pull/5338) due to incompatibilities in the Atlas search implementations.

### CODE REVIEW

File: libs/langchain/langchain/vectorstores/mongodb_atlas.py [hunk 1]
```diff
@@ -145,7 +159,32 @@ def _insert_texts(self, texts: List[str], metadatas: List[Dict[str, Any]]) -> Li
         insert_result = self._collection.insert_many(to_insert)  # type: ignore
         return insert_result.inserted_ids
 
-    def _similarity_search_with_score(
+    def _similarity_search_query(
+        self,
+        query: dict[str, Any],
```
Comment (blink1073) [inline libs/langchain/langchain/vectorstores/mongodb_atlas.py:164]: Nit: I suggest using the capitalized versions of `Dict` and `List` throughout for consistency.

Comment (NoahStapp) [reply libs/langchain/langchain/vectorstores/mongodb_atlas.py]: Done, thanks!

**Review by blink1073 — APPROVED** [review]
LGTM!

**Comment by baskaryan** [comment]
lgtm, thank you @NoahStapp!

**Comment by prakul** [comment]
This PR introduces support for $vectorSearch MQL stage which would be available with MongoDB v7.0.2. The users on the prior versions of MongoDB would be required to pin their langchain version to v0.0.304

**Pull request merged**